feat: Add a plot to ComparisonReport summarize display

MarieSacksick opened the pull request:
The plot of this display will be a pair plot to visualize one metric against another.

This is part of the narrative "I have several models, how can I choose the best one?". 

Closes #2129

MarieSacksick committed:
merge

MarieSacksick committed:
value error based on literal

MarieSacksick committed:
plot for comparison report for estimator

MarieSacksick committed:
chore: Change to more explicit class name and add docs

MarieSacksick committed:
Update sphinx/user_guide/reporters.rst

Co-authored-by: Guillaume Lemaitre <guillaume@probabl.ai>

MarieSacksick committed:
alphabetic sorting

MarieSacksick committed:
fix init for sphinx

MarieSacksick committed:
add function to be consistant with a display

MarieSacksick committed:
fix init for sphinx

MarieSacksick committed:
docs: explain a bit more about display and their functions

MarieSacksick committed:
linting

MarieSacksick committed:
introduce temporarily _SCORE_OR_LOSS_INFO in class

MarieSacksick committed:
first version of plot for comp report ready

MarieSacksick committed:
linting

MarieSacksick committed:
adapt to name change from report_metrics to summarize

MarieSacksick commented:
tests to come; ready for a first high-level review.

MarieSacksick committed:
add some tests

MarieSacksick committed:
add test not implemented error

MarieSacksick committed:
add tests

MarieSacksick committed:
add data_source at display creation

MarieSacksick committed:
add new tests

MarieSacksick committed:
add description to tests

MarieSacksick committed:
add example

glemaitre commented on skore/src/skore/sklearn/_plot/metrics/metrics_summary_display.py:
```diff
@@ -9,19 +13,166 @@ class MetricsSummaryDisplay(HelpDisplayMixin, StyleDisplayMixin):
+        # Add labels to the points with a small offset
+        text = self.summarize_data.columns.tolist()
+        for label, x_coord, y_coord in zip(text, x_data, y_data, strict=False):
+            self.ax_.annotate(
```
Looking at the rendering, we will want to have the option to put those as a legend outside of the plot instead of an annotation. It might a safer choice: one can avoid to show the legend and then tweak to have the annotation. Removing the annotation should be much more work.

glemaitre commented on skore/src/skore/sklearn/_plot/metrics/metrics_summary_display.py:
```diff
@@ -9,19 +13,166 @@ class MetricsSummaryDisplay(HelpDisplayMixin, StyleDisplayMixin):
+        if self.data_source is not None:
+            title += f" on {self.data_source} set"
+
+        self.ax_.scatter(x=x_data, y=y_data)
```
I think that by default we should use a set of marker and colors to plot each individual data point.

glemaitre commented on skore/src/skore/sklearn/_plot/metrics/metrics_summary_display.py:
```diff
@@ -9,19 +13,166 @@ class MetricsSummaryDisplay(HelpDisplayMixin, StyleDisplayMixin):
+                ),
+            )
+
+        plt.tight_layout()
```
You will not need this line.

glemaitre commented on skore/src/skore/sklearn/_plot/metrics/metrics_summary_display.py:
```diff
@@ -9,19 +13,166 @@ class MetricsSummaryDisplay(HelpDisplayMixin, StyleDisplayMixin):
+            title += f" on {self.data_source} set"
+
+        self.ax_.scatter(x=x_data, y=y_data)
+        self.ax_.set_title(title)
+        self.ax_.set_xlabel(x_label_text)
+        self.ax_.set_ylabel(y_label_text)
```
A more compact way to `set` stuff in matplotlib.

```suggestion
        self.ax_.set(title=title, xlabel=x_label_text, y_label=y_label_text)
```

glemaitre reviewed (commented):
It is a quick review. A couple of additional thoughts:

- we will want to have some logspace sometimes. In scikit-learn, we came with an heuristic with the time for the validation curve: https://github.com/scikit-learn/scikit-learn/blob/d4d4af8c4/sklearn/model_selection/_plot.py#L111-L117 I think that we can retake this implementation
- we need to put some thoughts regarding score vs. loss.
- I think that we need to tackle the following issue first: https://github.com/probabl-ai/skore/issues/1837

MarieSacksick committed:
Update skore/src/skore/sklearn/_plot/metrics/metrics_summary_display.py

Co-authored-by: Guillaume Lemaitre <guillaume@probabl.ai>

MarieSacksick committed:
change not implemented error for estimator

MarieSacksick committed:
bugfix matplotlib ax set

MarieSacksick committed:
remove useless line

MarieSacksick commented on skore/src/skore/sklearn/_plot/metrics/metrics_summary_display.py:
```diff
@@ -9,19 +13,166 @@ class MetricsSummaryDisplay(HelpDisplayMixin, StyleDisplayMixin):
     """
 
-    def __init__(self, summarize_data):
+    # should be removed once transformed into a utils
```
I don't really get why we would put this code into the MetricsAccessor, as it's only used in the display, and even only used in the plot of the display.  
Do you have something else in mind?

MarieSacksick commented on skore/src/skore/sklearn/_plot/metrics/metrics_summary_display.py:
```diff
@@ -9,19 +13,166 @@ class MetricsSummaryDisplay(HelpDisplayMixin, StyleDisplayMixin):
+        x_label = self._SCORE_OR_LOSS_INFO.get(x, {}).get("name", x)
+        y_label = self._SCORE_OR_LOSS_INFO.get(y, {}).get("name", y)
+
+        # Check that the metrics are in the report
```
Not exactly: for some plots, we will not have a y axis.  
I'd argue that we should do refactoring when we have the new code, and not before, as we don't know exactly what we will have in it.

MarieSacksick committed:
change annotation to legend

MarieSacksick commented:
> we will want to have some logspace sometimes. In scikit-learn, we came with an heuristic with the time for the validation curve: https://github.com/scikit-learn/scikit-learn/blob/d4d4af8c4/sklearn/model_selection/_plot.py#L111-L117 I think that we can retake this implementation  

About this, it looks like an improvement to add this option imo. It could be a good first issue. Since we have a workshop soon, what keeping it for this workshop?  

> I think that we need to tackle the following issue first: https://github.com/probabl-ai/skore/issues/1837

If you think that this other issue can be done soon, let's wait indeed! Otherwise, I don't really like keeping PR open for a while, we tend to forget the code and the context.  

> we need to put some thoughts regarding score vs. loss.  

Do you have the pareto front in mind?  
As the first feedback, to me it's an improvement, not light this one, and I prefer to do step by step and releasing this first part when we are ready.

glemaitre commented:
> About this, it looks like an improvement to add this option imo. It could be a good first issue. Since we have a workshop soon, what keeping it for this workshop?

It is 5 lines of code already tested somewhere else. I would not bother since I already know that we want it.

> Do you have the pareto front in mind?
As the first feedback, to me it's an improvement, not light this one, and I prefer to do step by step and releasing this first part when we are ready.

Not really, it was more on the direction of the loss/score. But I think we can just align the implementation with the display on the hub.

In terms of implementation, I don't like a partial implementation.

MarieSacksick committed:
Merge branch 'main' into plot_comp_report_metrics

MarieSacksick committed:
linting

mrastgoo committed:
docs(skore): Change the name of features after preprocessing (#1901)

closes #1421 

Created a separated dataframe for features importances and rename the
features by string replace operations.
Changes were made for two figures:  `engineered_ridge_report`

MarieSacksick committed:
finish merge

MarieSacksick committed:
annot to legend

MarieSacksick committed:
Merge branch 'main' into plot_comp_report_metrics

MarieSacksick committed:
change scale according to data range

MarieSacksick commented:
@glemaitre, can you be more precise about your expectations about the following sentence please?

> Not really, it was more on the direction of the loss/score. But I think we can just align the implementation with the display on the hub.

So far, the only difference with the hub is the display of the pareto front and the colored aspect on the "worse" side of the front.

<img width="739" height="358" alt="image" src="https://github.com/user-attachments/assets/6b01706e-9649-4e6b-903d-3a5ad73e0021" />

MarieSacksick committed:
extend to support custom metrics

MarieSacksick committed:
fix: remove useless subplot

MarieSacksick committed:
add test for axis

MarieSacksick committed:
fix test

MarieSacksick committed:
Merge branch 'main' into plot_comp_report_metrics

MarieSacksick committed:
Merge branch 'main' into plot_comp_report_metrics

MarieSacksick committed:
linting and update import to be consistant with refactor

MarieSacksick committed:
add default plot function in metrics summary display

MarieSacksick committed:
fix docs with available metrics

MarieSacksick committed:
Merge branch 'main' into plot_comp_report_metrics
